asdf/asdftypes.py

```diff
@@ -696,7 +703,27 @@ def to_tree_tagged(cls, node, ctx):
             An instance of `asdf.tagged.Tagged`.
         """
         obj = cls.to_tree(node, ctx)
-        return tagged.tag_object(cls.yaml_tag, obj, ctx=ctx)
+        yaml_tag = cls.yaml_tag
+
+        node_cls = type(node)
+
+        if node_cls.__name__ in cls._subclass_map and isinstance(obj, dict):
+            obj['subclass'] = node_cls.__name__
```

Comment:
While this works, to me it seems brittle and inelegant. Ideally we could build this into the standard itself somehow. 

One way to do this would be to use URI fragments, which are allowed by the YAML standard. For example, if the base tag is `tag:nowhere.org/custom/foo-1.0.0`, then a subclass could be tagged with something like `tag:nowhere.org/custom/foo-1.0.0#subclass=FooSubclass`.

This would require a few changes to the validation machinery and also to reference resolution, but it might be worth the effort. The main question is how other ASDF implementations would handle this. Also it would be important to make sure that subclasses are namespaced properly.